lib/manageiq/providers/ibm_cloud/cloud_tools/authentication.rb

```diff
@@ -0,0 +1,138 @@
+
+              BearerAuth.new(bearer_info, :logger => logger)
+            end
+
+            # Fetch a new Authenticator using IAM token.
+            # @param api_key [String] The IAM OAuth api token.
+            # @param bearer_info [Hash{Symbol => String, Integer}] bearer hash with token, expire_time and api_key as keys.
+            # @param logger [Logger, IO, String] Either a Logger object, IO object or path to file.
+            #
+            # @return [BearerAuth]
+            def new_auth(api_key: nil, bearer_info: nil, logger: nil)
+              raise 'No authentication information given.' if api_key.nil? && bearer_info.nil?
+
+              bearer_info[:api_key] = api_key if !bearer_info.nil? && !api_key.nil?
+
+              return new_bearer(bearer_info, :logger => logger) unless bearer_info.nil?
+
+              iam_auth = new_iam(api_key, :logger => logger)
+              new_bearer(iam_auth.bearer_info)
+            end
+          end
+
+          # API Authentication using Bearer header.
+          # @see https://github.com/IBM/ruby-sdk-core/blob/main/lib/ibm_cloud_sdk_core/authenticators/bearer_token_authenticator.rb SDK Bearer doc.
+          class BearerAuth < IbmVpc::Authenticators::BearerTokenAuthenticator
+            # Convert bearer info into something the superclass understands.
+            # @param bearer_info [Hash{Symbol=>String, Integer}] bearer hash with token, expire_time and api_key as keys.
+            # @param logger [Logger, IO,String] Either a logger object, a file path or IO object.
+            #
+            # @return [void]
+            def initialize(bearer_info, logger: nil)
+              @logger = define_logger(logger)
+              @bearer_info = verify_info(bearer_info)
+              super({:bearer_token => bearer_info[:token]})
+            end
+
+            # @return [Hash{Symbol=>String, Integer}] bearer hash with token, expire_time and api_key as keys.
+            attr_reader :bearer_info
+
+            private
+
+            # Verify that the bearer token hasn't expired. If it has and an API Key is present try to get a new bearer token.
+            #
+            # @return [Hash{Symbol => String, Integer}] @see #bearer_info
+            def verify_info(bearer_info)
+              # Raise standard error if expiration time expires in the next 10 second.
+              if verify_valid(bearer_info[:expire_time])
+                raise 'Bearer token has expired.' if bearer_info[:api_key].nil?
+
+                @logger.info('Bearer token expired. Fetching new one using API Key.')
+                return IamAuth.new(bearer_info[:api_key]).bearer_info
+              end
+              bearer_info
+            end
+
+            # Validate expire time is set and greater than now - 10 seconds.
+            #
+            # @return [Boolean]
+            def verify_valid(expire_time)
+              return false if expire_time.nil?
+
+              expire_time <= (Time.now.to_i - 10)
```

Comment:
Wouldn't the token be valid if the `expire_time > Time.now.to_i` or `expire_time > Time.now.to_i + 10` if you want to give a little "buffer" when it is still valid but you want people to renew it?

Reply:
Oh you're expecting `true` returned from `verify_valid` to mean it has expired?  That's a bit counterintuitive, maybe rename this method `#expired?`

I would expect a truthy return from a method called `valid` would mean it was valid not invalid

Reply:
The name of the method should be 'is_expired'. Expired is true, Valid is false.

Reply:
Yeah, that got me too. I must have named the method before finalizing the logic and forgot to update the method name.